rust/core-lib/src/event_context.rs

```diff
@@ -137,8 +161,37 @@ impl<'a> EventContext<'a> {
                 }),
             SpecialEvent::RequestLines(LineRange { first, last }) =>
                 self.do_request_lines(first as usize, last as usize),
-            SpecialEvent::RequestHover{ request_id, position } =>
-                self.do_request_hover(request_id, position)
+            SpecialEvent::RequestHover { request_id, position } =>
+                self.do_request_hover(request_id, position),
+            SpecialEvent::ToggleRecording(_) => {}
+            SpecialEvent::PlayRecording(recording_name) => {
+                let recorder = self.recorder.borrow();
+
+                // Don't group with the previous action
+                self.editor.borrow_mut().update_edit_type();
+                self.editor.borrow_mut().calculate_undo_group();
+
+                // No matter what, our entire block must belong to the same undo group
+                self.editor.borrow_mut().set_force_undo_group(true);
+                recorder.play(&recording_name, |event| {
+                    self.broadcast_event(event.clone());
+
+                    let mut editor = self.editor.borrow_mut();
+                    let (delta, last_text, keep_sels) = match editor.commit_delta() {
+                        Some(edit_info) => edit_info,
+                        None => return,
+                    };
+                    self.update_views(&editor, &delta, &last_text, keep_sels);
```

Comment:
This is interesting.

Basically, a consequence of our approach here is that we aren't properly notifying plugins of these edits; after `commit_delta` we're supposed to ship the new delta off to plugins, and we're skipping that step.

I can think of one maybe simple solution; before playback, get the current head revision from the engine, and then _after_ playback compute the delta from that revision to the current head (`Engine::delta_rev_head`) and _then_ send that to the plugins as per `after_edit`. I'm not _sure_ this will work, and we're definitely getting into uncomfortably hacky territory, but maybe start there and see what you figure out?